rust/src/treefile.rs

```diff
@@ -160,6 +161,31 @@ fn load_passwd_file<P: AsRef<Path>>(
     return Ok(None);
 }
 
+/// Look for use of legacy/renamed fields and migrate them to the new field.
+fn migrate_legacy_fields(treefile: &mut TreeComposeConfig) -> Fallible<()> {
+    macro_rules! migrate_field {
+        ( $legacy:ident, $new:ident ) => {{
+            if treefile.$legacy.is_some() && treefile.$new.is_some() {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    format!("Cannot use new and legacy forms of {}", stringify!($new)),
+                ).into());
+            }
+            if treefile.$legacy.is_some() {
+                treefile.$new = treefile.$legacy.take();
+            }
```

Comment:
There's a nicer version we could use here:

```rust
treefile.$new = treefile.$new.or(treefile.$legacy.take());
```

But [or](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.or) consumes the `Option`, while we're borrowing here. And for some reason I felt a bit odd calling this function as:

```rust
treefile = migrate_legacy_fields(treefile);
```

Reply:
Wouldn't it be cleaner as a method on the treefile that takes `Self` i.e.

```
   fn migrate_legacy_fields(Self) -> Self { ... }
```